File: backend/functions/tenantSignup.js
const bcrypt = require('bcryptjs');
const { DynamoDBClient } = require('@aws-sdk/client-dynamodb');
const { DynamoDBDocumentClient, PutCommand } = require('@aws-sdk/lib-dynamodb');
const { v4: uuidv4 } = require('uuid');

const dynamoClient = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(dynamoClient);

exports.handler = async (event) => {
    try {
        const { email, password, companyName, firstName, lastName } = JSON.parse(event.body);
        
        // Generate tenant ID
        const tenantId = uuidv4();
        
        // Hash password
        const hashedPassword = await bcrypt.hash(password, 10);
        
        // Create tenant record
        const tenantParams = {
            TableName: process.env.TENANTS_TABLE,
            Item: {
                tenantId,
                companyName,
                createdAt: new Date().toISOString(),
                status: 'active'
            }
        };
        
        // Create admin user record
        const userParams = {
            TableName: process.env.USERS_TABLE,
            Item: {
                email,
                password: hashedPassword,
                tenantId,
                firstName,
                lastName,
                role: 'admin',
                createdAt: new Date().toISOString()
            }
        };
        
        // Save both records
        await Promise.all([
            docClient.send(new PutCommand(tenantParams)),
            docClient.send(new PutCommand(userParams))
        ]);
        
        return {
            statusCode: 201,
            body: JSON.stringify({
                message: 'Tenant created successfully',
                tenantId
            })
        };
    } catch (error) {
        console.error('Tenant signup error:', error);
        return {
            statusCode: 500,
            body: JSON.stringify({ message: 'Internal server error' })
        };
    }
}; 

File: backend/functions/tenant_signup.py
import json
import os
import uuid
import bcrypt
import boto3
from datetime import datetime

dynamodb = boto3.resource('dynamodb')
users_table = dynamodb.Table(os.environ['USERS_TABLE'])
tenants_table = dynamodb.Table(os.environ['TENANTS_TABLE'])

def lambda_handler(event, context):
    try:
        body = json.loads(event['body'])
        email = body['email']
        password = body['password']
        company_name = body['companyName']
        first_name = body['firstName']
        last_name = body['lastName']
        
        # Generate tenant ID
        tenant_id = str(uuid.uuid4())
        
        # Hash password
        hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
        
        # Create tenant record
        tenant_item = {
            'tenantId': tenant_id,
            'companyName': company_name,
            'createdAt': datetime.utcnow().isoformat(),
            'status': 'active'
        }
        
        # Create admin user record
        user_item = {
            'email': email,
            'password': hashed_password,
            'tenantId': tenant_id,
            'firstName': first_name,
            'lastName': last_name,
            'role': 'admin',
            'createdAt': datetime.utcnow().isoformat()
        }
        
        # Save both records
        with dynamodb.batch_writer() as batch:
            batch.put_item(
                TableName=os.environ['TENANTS_TABLE'],
                Item=tenant_item
            )
            batch.put_item(
                TableName=os.environ['USERS_TABLE'],
                Item=user_item
            )
        
        return {
            'statusCode': 201,
            'body': json.dumps({
                'message': 'Tenant created successfully',
                'tenantId': tenant_id
            })
        }
    except Exception as e:
        print(f'Tenant signup error: {str(e)}')
        return {
            'statusCode': 500,
            'body': json.dumps({'message': 'Internal server error'})
        } 